### ISSUE TITLE
BUILDING.md: sharing info on direct build dependency list

### ISSUE THREAD

**Issue by Anton-Latukha** [projectM-visualizer/projectm#246]
I've just successfully `2.1.0 -> 3.1.0` `projectM` on NixOS:
https://github.com/NixOS/nixpkgs/pull/70433

If you interested, minimal direct build dependencies for `--emable-qt --enable-sdl` are:
```
, autoconf
, automake
, pkgconfig
, SDL2
, qtdeclarative
, libpulseaudio
, glm
```

_Some of them with `dev` parts probably, that is automated in tooling._

_That `libtool` can be omitted - it is almost a sin doing on my part._

1. But, if SDL2, the build does not need SDL v1.

2. Personally, I did not found an option for Jack, maybe it is fully autodetects?

3. Default `qt5` seems unneeded, 1) it is indirect dep of `qtdeclarative`, 2) indirect deps are ommited from builds in Nix tooling, so I can assure you that project builds with operational `qtdeclarative` provided alone.

Thank you for everything.

**Comment by Anton-Latukha** [comment]
Ok. I've just seen that Jack seems like has hooks in the configuration system, and relies on unmaintained branches, but it also does not declared as an option in configure information or output.

**Comment by revmischa** [comment]
PRs are extremely welcome!
1. I don't think anything should be using sdl1 at this point, if it is we should fix that

2. jack is a configure option? `AC_ARG_ENABLE([jack], AS_HELP_STRING([--enable-jack], [Build Jack]), [], [enable_jack=check])` - it's in the configure --help
```
  --enable-qt             Enable Qt: needed for pulseaudio and jack GUIs
  --enable-pulseaudio     Build Pulseaudio
  --enable-jack           Build Jack
```

**Comment by Anton-Latukha** [comment]
My bad on Jack, thank you.

**Comment by Anton-Latukha** [comment]
1. I thought it is too hardcore for me to come and change the main build docs.
2. While I am providing solid info from my experience and expertise, I do not know the project from inside-out, and all the applications the people are using project in, 
3. so it would be clumsy of me & without a tact to change the main dev docs right away &
4. if I do that - that would generally result in more back and forth written messages and several versions of the patch, several timeouts and comebacks, several checks & take more time for everyone.

:arrow_up: That is probably the algorythm that reasonable contributor follows, & why I sometimes prefer to write report, rather then to send a patch.

Also to right away approve, inform and ask to send patch is also right.

Ok. I would open a proposal.

**Comment by revmischa** [comment]
The docs are incomplete and outdated, so any attempt to correct and update them is greatly appreciated

**Comment by Anton-Latukha** [comment]
<details><summary>The thing about Jack is:</summary>
<p>

```fish
/configure --help
`configure' configures projectM 3.1.0-beta1 to adapt to many kinds of systems.

Usage: ./configure [OPTION]... [VAR=VALUE]...

To assign environment variables (e.g., CC, CFLAGS...), specify them as
VAR=VALUE.  See below for descriptions of some of the useful variables.

Defaults for the options are specified in brackets.

Configuration:
  -h, --help              display this help and exit
      --help=short        display options specific to this package
      --help=recursive    display the short help of all the included packages
  -V, --version           display version information and exit
  -q, --quiet, --silent   do not print `checking ...' messages
      --cache-file=FILE   cache test results in FILE [disabled]
  -C, --config-cache      alias for `--cache-file=config.cache'
  -n, --no-create         do not create output files
      --srcdir=DIR        find the sources in DIR [configure dir or `..']

Installation directories:
  --prefix=PREFIX         install architecture-independent files in PREFIX
                          [/usr/local]
  --exec-prefix=EPREFIX   install architecture-dependent files in EPREFIX
                          [PREFIX]

By default, `make install' will install all the files in
`/usr/local/bin', `/usr/local/lib' etc.  You can specify
an installation prefix other than `/usr/local' using `--prefix',
for instance `--prefix=$HOME'.

For better control, use the options below.

Fine tuning of the installation directories:
  --bindir=DIR            user executables [EPREFIX/bin]
  --sbindir=DIR           system admin executables [EPREFIX/sbin]
  --libexecdir=DIR        program executables [EPREFIX/libexec]
  --sysconfdir=DIR        read-only single-machine data [PREFIX/etc]
  --sharedstatedir=DIR    modifiable architecture-independent data [PREFIX/com]
  --localstatedir=DIR     modifiable single-machine data [PREFIX/var]
  --libdir=DIR            object code libraries [EPREFIX/lib]
  --includedir=DIR        C header files [PREFIX/include]
  --oldincludedir=DIR     C header files for non-gcc [/usr/include]
  --datarootdir=DIR       read-only arch.-independent data root [PREFIX/share]
  --datadir=DIR           read-only architecture-independent data [DATAROOTDIR]
  --infodir=DIR           info documentation [DATAROOTDIR/info]
  --localedir=DIR         locale-dependent data [DATAROOTDIR/locale]
  --mandir=DIR            man documentation [DATAROOTDIR/man]
  --docdir=DIR            documentation root [DATAROOTDIR/doc/projectM]
  --htmldir=DIR           html documentation [DOCDIR]
  --dvidir=DIR            dvi documentation [DOCDIR]
  --pdfdir=DIR            pdf documentation [DOCDIR]
  --psdir=DIR             ps documentation [DOCDIR]

Program names:
  --program-prefix=PREFIX            prepend PREFIX to installed program names
  --program-suffix=SUFFIX            append SUFFIX to installed program names
  --program-transform-name=PROGRAM   run sed PROGRAM on installed program names

System types:
  --build=BUILD     configure for building on BUILD [guessed]
  --host=HOST       cross-compile to build programs to run on HOST [BUILD]

Optional Features:
  --disable-option-checking  ignore unrecognized --enable/--with options
  --disable-FEATURE       do not include FEATURE (same as --enable-FEATURE=no)
  --enable-FEATURE[=ARG]  include FEATURE [ARG=yes]
  --enable-silent-rules   less verbose build output (undo: "make V=1")
  --disable-silent-rules  verbose build output (undo: "make V=0")
  --enable-dependency-tracking
                          do not reject slow dependency extractors
  --disable-dependency-tracking
                          speeds up one-time build
  --enable-shared[=PKGS]  build shared libraries [default=yes]
  --enable-static[=PKGS]  build static libraries [default=yes]
  --enable-fast-install[=PKGS]
                          optimize for fast installation [default=yes]
  --disable-libtool-lock  avoid locking (might break parallel builds)
  --disable-rpath         patch libtool to not use rpath
  --enable-emscripten     Build for web with emscripten
  --enable-sdl            Build SDL2 app
  --disable-sdltest       Do not try to compile and run a test SDL program
  --disable-sdlframework Do not search for SDL2.framework
  --enable-threading      multhreading
  --enable-gles           OpenGL ES support
  --enable-qt             Build Qt

Optional Packages:
  --with-PACKAGE[=ARG]    use PACKAGE [ARG=yes]
  --without-PACKAGE       do not use PACKAGE (same as --with-PACKAGE=no)
  --with-pic[=PKGS]       try to use only PIC/non-PIC objects [default=use
                          both]
  --with-aix-soname=aix|svr4|both
                          shared library versioning (aka "SONAME") variant to
                          provide on AIX, [default=aix].
  --with-gnu-ld           assume the C compiler uses GNU ld [default=no]
  --with-sysroot[=DIR]    Search for dependent libraries within DIR (or the
                          compiler's sysroot if not specified).
  --with-xquartz-gl[=DIR] On Mac OSX, use opengl provided by X11/XQuartz
                          instead of the built-in framework. If enabled, the
                          default location is [DIR=/opt/X11]. This option is
                          default to false.
  --with-sdl-prefix=PFX   Prefix where SDL is installed (optional)
  --with-sdl-exec-prefix=PFX Exec prefix where SDL is installed (optional)

Some influential environment variables:
  CC          C compiler command
  CFLAGS      C compiler flags
  LDFLAGS     linker flags, e.g. -L<lib dir> if you have libraries in a
              nonstandard directory <lib dir>
  LIBS        libraries to pass to the linker, e.g. -l<library>
  CPPFLAGS    (Objective) C/C++ preprocessor flags, e.g. -I<include dir> if
              you have headers in a nonstandard directory <include dir>
  LT_SYS_LIBRARY_PATH
              User-defined run-time library search path.
  CPP         C preprocessor
  CXX         C++ compiler command
  CXXFLAGS    C++ compiler flags
  CXXCPP      C++ preprocessor
  PKG_CONFIG  path to pkg-config utility
  PKG_CONFIG_PATH
              directories to add to pkg-config's search path
  PKG_CONFIG_LIBDIR
              path overriding pkg-config's built-in search path
  GL_CFLAGS   C compiler flags for GL, overriding configure script defaults
  GL_LIBS     Linker flags for GL, overriding configure script defaults
  SDL2_FRAMEWORK
              Path to SDL2.framework
  SDL_CFLAGS  C compiler flags for SDL, overriding pkg-config
  SDL_LIBS    linker flags for SDL, overriding pkg-config
  QT_CFLAGS   C compiler flags for QT, overriding pkg-config
  QT_LIBS     linker flags for QT, overriding pkg-config
  LIBPULSE_CFLAGS
              C compiler flags for LIBPULSE, overriding pkg-config
  LIBPULSE_LIBS
              linker flags for LIBPULSE, overriding pkg-config

Use these variables to override the choices made by `configure' or to help
it to find libraries and programs with nonstandard names/locations.

Report bugs to <mischa@mvstg.biz>.
```

</details>

Or, in other words:
```fish
./configure --help | grep -i jack
(nothing)
```

I also can not figure-out, seems like `Qt` beaing "still" required for Jack build, and I can not understand if Jack1 or Jack2, and how it builds somehow.

Oh:
```fish
./configure --enable-jack | grep -i jack
configure: WARNING: unrecognized options: --enable-jack
```

_if something, this is state of `3.1.0`._

**Comment by revmischa** [comment]
Can you please try building from source or the latest 3.1.1-RC4? https://github.com/projectM-visualizer/projectm/releases

**Comment by Anton-Latukha** [comment]
Note: seeing you `v3.1.1-rc3` -> `3.1.1-rc4`, dropped `v` in the git tag, wich I am keen to. This tags are a source of distribution automation for detecting upstream release and triggering packaging pipeline (at lease in our current tooling), so, in my case that `v` change should be tracked by me so bot would find your new releases & would try roll autoupdate & CI build & test and in success - roll-out your new releases into distribution.

**Comment by Anton-Latukha** [comment]
I have an interest.

How SDL connects to the Linux sound system? Does it use/listen the default ALSA audio sink/device by default?

**Comment by revmischa** [comment]
Quite right - try v3.1.1-rc5 :)
I'm not sure what SDL uses by default on Linux - ProjectM is using [OpenAudioDevice](https://wiki.libsdl.org/SDL_OpenAudioDevice?highlight=%28%5CbCategoryAudio%5Cb%29%7C%28CategoryEnum%29%7C%28CategoryStruct%29)

If you're in Kyiv we can grab a 🍺

**Comment by Anton-Latukha** [comment]
I am just interested, since everything still in the end outputs using ALSA. Maybe only now PipeWire can somewhere in the lab output into Kernel without ALSA by doing its emulation.

I am just interested, if PulseAudio on 99% of setups loops on itself the default ALSA output device, outputting sound back into ALSA hardware device to feed the hardware.

If SDL hooks to the default ALSA device - that means it supports... oh no, I now saw the flaw in my logic. If PulseAudio recieves and sends to low-level ALSA device - SDL does not have a clue or hook-in in what/in device in the system configuration.

_P.S._
_... & still I am salty on ALSA and their internal design war and exitus of good people before 1.0_

**Comment by Anton-Latukha** [comment]
Not direcly in Kyiv currently, lived there for 12 years, now in Bila Tserkva, 80km near, at the parents house. Kind of regrouping, in search of fitting position/work.

On the occasion/invitation I can travel and we can meet thou.

**Comment by coast77777777777** [comment]
I think we need Linux equivalents of WASAPI. the sdl build needs loopback for non-windows builds to visualize whatever the operating system is playing over the speakers. Does this seem correct? Did valve already handle this is SDL?

**Comment by revmischa** [comment]
I really don't know how such a thing would work - I assumed the pulseaudio or jack version could do something like this. SDL is supposed to be platform-agnostic and so I'm not sure how much you can get into these sort of details with it. 
If you are using some system which does allow you to route audio internally and exposes it as an audio device (like Soundflower on macOS) then it will work with the SDL app. If not, other solutions may be better, for example the iTunes plugin which I've been maintaining.

**Comment by Anton-Latukha** [comment]
AFAIK the ALSA, not only seem to be the only real low-level current Linux subsystem, but also is **the** holder of all Linux Kernel sound drivers.

I mean, literally, the first most mainstream text we read about ALSA:
> Advanced Linux Sound Architecture (ALSA) is a software framework and part of the Linux kernel that provides an application programming interface (API) for sound card device drivers.

Also, there is a well-known in the Linux Kernel community, the main developer of the Linux Kernel, titan that holds all ALSA and it's drivers on his shoulders mostly alone for last decades. His name is Takashi Iwai. If there was a real Linux sound question or plea for help somewhere - he was the only man people was sending to, and only that solved all that. I almost pressed a send button to send message written to him, because of my pain about Asus Xonar DS, but then I desided that I do not want to bother him whith what I can live with.

He written a huge portion of drivers that everyone uses. In last two years the new friend & titan joined him, now they are togather.

====

I mean by that, how on Earth the SDL can do sound input/output without the Linux Kernel drivers, which are ALSA. Only maybe by having inside the project own user-space compiled-in ALSA that somehow works with Kernel through user-space API (there is no module required for SDL).
In other words, it indeed uses ALSA.

**Comment by revmischa** [comment]
I think the other cross-platform option is [PortAudio](http://portaudio.com/docs/v19-doxydocs/api_overview.html)

**Comment by Anton-Latukha** [comment]
I am perfectly happy with SDL, I just asked how it hooks-up to Linux sound system.
We are talking good here, but we think from different perspectives. Since I am maintainer I am thinking from down-up how into Linux system sound stream the `projectM` connects, and since you are `projectM` devs, you think from `projectM` down, how `projectM` is developed and operates.
I am very happy that `projectM` uses SDL, which utilizes OpenAL, & I love OpenAL very mutch.

**Comment by coast77777777777** [comment]
We require loopback. Can ALSA community show how for all kernels?

**Comment by Anton-Latukha** [comment]
I am myself not really an ALSA community guy. I just was following ALSA thing when prehistoric ALSA drama happened, because 0.9 ALSA already was useble and used in distros.

Then in years I worked and played quite a bit with ALSA. I am not a C or C++ dev. If I am a community member - it is a Spacemacs & Haskell & NixOS communities.

I do not know for what an ALSA dev needed here for.

Maybe what you really need is some simple ALSA configuration. ALSA has puzzlingly cryptic but pretty flexible configuration DSL lang inside. Adding a sink or muxer that sends input into two outputs, one onto hardware, the other goes back. Something like that, or backwards - for backward loopback. Or to config a sink so your interface receives the copy of default input or output. It of, course, all works for all ALSAs and Kernels and whatever uses ALSAs hardware devices not direcly (almost nothing directly uses ALSA hardware devices), it always goes through the mixer in ALSA.

The loopback also can mean some other loop-backing. Like, with a time-shift, or something in SDL.

**Comment by Anton-Latukha** [comment]
BTW, about no-software directly using ALSA devices. You can easily seize a device and do a bit-perfect output directly to hardware sound device, for example in DeadBeef, SMPlayer, VLC, or any other player that exposes the list of raw devices. The sound cards mostly have no mixers on them, hardware works direcly - accept one input stream and DACs it. Moslty PulseAudio seizes that one output stream to device, So when you output through ALSA, you must by quality kill PulseAudio, to connect the player to the sound device. You can even set or send any supported by deveice Bit-wides (8, 16, 24, 32) and supported KHz directly into it.

There is also a possibility to monitor ALSA device states, stream parameters, and what is connected - just look into `/proc`.

**Comment by Anton-Latukha** [comment]
I just happen to have good telco education, I can talk on audio transmission, receiving & encoding for hours. And I love Linux and music.

**Comment by coast77777777777** [comment]
Thanks for sharing all the good information. Let’s keep the conversation going. I added supported for windows loopback a while back to projectm. Before, the software required a microphone aka physical audio capture device. This source audio capture device provides the source audio signal that results in projectm visuals. I accomplished this operating system audio loopback capture using WASAPI library from Microsoft. For example, I can open any video on YouTube and then open projectm and the visuals synchronize.  The issue with Linux builds of projectm is that I again need a microphone to capture the audio. I have not invested the time to resolve operating system audio loopback capture for Linux. That is what’s needed. What do you think?

**Comment by Anton-Latukha** [comment]
@coast77777777777 

As I said, it can be done with some simple configuration. I would sometime refresh on all that, think/search how to do that, maybe there is simple way to attach.

Also currently I found this solution:
https://blog.getreu.net/projects/snd-aloop-device/body.html - is this is the right feature?

It is a loopback. But it is not a splitter, when PA, physics of sound medium, and microphone compose a 1-many-splitter to with is connected one 1-1-loopback.

This shared article is a component of config I am talking about, it is a loopback device.

For splitter it is what I described above in strict terms the good answer is here:
https://superuser.com/questions/1248439/can-i-pipe-input-from-a-linux-soundcard-into-2-different-applications

_(in device here I mean internal ALSA term. "Device - sound structure entity")_
It is to create the `plughw` device before the any device you want to snoop on. `plughw` enables to have `dsnoop` device on it.

**Comment by Anton-Latukha** [comment]
Documentation of ALSA features and configuration language become documented and much-much better: https://wiki.archlinux.org/index.php/Advanced_Linux_Sound_Architecture#Basic_syntax

I used to learn about those things by days of try and error.

**Comment by Anton-Latukha** [comment]
Also for spliter & loopback - there is even very simple but crude poor-mans version:
```sh
commandOutput | tee output1 output2
```
Like this fun stuff:
https://debian-administration.org/article/145/use_and_abuse_of_pipes_with_audio_data

I think `tee` is POSIX, it is in `coreutils`.

**Comment by Anton-Latukha** [comment]
In Linux - you can even just do: `command | tee output1 > /dev/snd/pcmC0D0p` (or any other which you like there) - which would directly send sound to play on the device, the device must be free thou.

**Comment by Anton-Latukha** [comment]
Ok, the head post topic ends its meaning here. I shared the info, and improved docs, and what is left from head post is now a very minor question. I am closing this report.

We had a good talk here, and I answered several questions raised in it.

If there would be further requests/questions, write to me on my emails or create a report and call me into.

Thank you for the discussion.

**Comment by coast77777777777** [comment]
I appreciate the information and general direction
